Make self-driving docs bolder and more actionable (#18369)
PostHog/posthog.com: contents/docs/self-driving/anatomy-of-a-pr.mdx

---
title: Anatomy of a pull request
sidebarTitle: Anatomy of a PR
sidebar: Docs
showTitle: true
---

import { CallToAction } from 'components/CallToAction'

<CalloutBox icon="IconFlask" title="Open beta" type="action">

Self-driving is in open beta. It's improving quickly – expect rough edges, and expect them to disappear fast.

</CalloutBox>

> The rest of these docs describe the [self-improving loop](/docs/self-driving/self-improving-loop) in the abstract. This page follows one bug all the way through it, from the first signal to the measured fix. The product is a composite example, but every step is exactly what self-driving does.

Reviewer: ```suggestion
The concepts docs describe the [self-improving loop](/docs/self-driving/self-improving-loop) in the abstract. This page follows one bug all the way through it, from the first signal to the measured fix. The product is a composite example, but every step is exactly what self-driving does.
```

Reviewer: "rest of these" is a little confusing and makes me think: this page? all the docs? this is a bit more specific

Author: Applied verbatim – the opener now points at the concepts docs instead of "the rest of these docs."

One problem rarely shows up as one clean signal, so the pipeline deduplicates the stream and clusters the signals pointing at the same underlying issue into a single [report](/docs/self-driving/reports):

- **Title:** Checkout confirm fails when the cart contains a discounted item
- **Grouped signals:** the exception, the rage-click cluster, and both support tickets
- **Priority:** P1 – it's in a hot path (checkout), it's breaking the product for affected users, and the affected users include paying customers
- **State:** to be determined by investigation

Instead of a dozen alerts across three tools, your [inbox](/docs/self-driving/inbox) gets one framed problem with the evidence attached.

## 3. An agent investigates

An agent picks up the report and digs into both sides of the story:

- **Your PostHog data** – when the error started, which release it correlates with, how many users and how much revenue sit behind it.
- **Your codebase** – it pairs the stack trace with the code and finds the root cause: `applyDiscount()` returns `undefined` for carts using a legacy coupon format, and the confirm step reads `.total` off the result without checking.

> The verdict: the problem is real, the fix is concrete, and the report is **actionable**. If it had needed a product decision instead – say, two valid ways to handle legacy coupons – it would surface in your inbox marked **needs input** and wait for you, for free.

Reviewer: ```suggestion
The verdict: the problem is real, the fix is concrete, and the report is marked **actionable**. If it had needed a product decision instead – say, two valid ways to handle legacy coupons – it would surface in your inbox marked **needs input** and wait for you, for free.
```

Author: Applied – "marked **actionable**" it is.

4. It suggests a reviewer from git blame, and babysits CI until the checks are green.

The PR reads like it came from a teammate who did the investigation properly, because that's what happened.

## 5. You review and merge

The PR lands in your inbox ranked with everything else. You review it the way you'd review anyone's code: read the diff, check the test, ask for changes if something's off. The agent responds to review feedback and updates the branch.

Nothing ships without you. This is also the moment you're billed: a flat $15 for the pull request, and [if it wasn't worth paying for, we refund it](/docs/self-driving/pricing#refunds).

## 6. PostHog measures the result

After the fix ships, the loop closes:

- Checkout conversion recovers to its baseline, and the exception stops firing.
- After a soak window, a validation scout re-checks that the fix actually held, instead of assuming the merge was the end of the story.
- The outcome feeds back in as new signals, so the next pass works from better data than the last.

## The whole loop, on one bug

> From first exception to merged, verified fix: about a day. Along the way, nobody filed a ticket, triaged an alert channel, or reconstructed the bug from a support thread. That's the difference between monitoring, which tells you something broke, and self-driving, which hands you the fix.

Reviewer: in the set up at the beginning of the doc, we do say: "Zendesk files two fresh tickets the next morning, both variations of "I can't complete my order." 

so "nobody filed a ticket" feels a little misleading. tickets are still a valuable signal source, the goal is just to reduce them or triage the issue before you are flooded with them

Author: Good catch – reworded in 28ae1b6 so the tickets still arrive but become evidence in a report that's already moving, rather than a queue to triage.
